File: kakao.app/js/ked/view/colmns/dirtree.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.prototype.hasOwnProperty(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, lpad: function (l,s='',c=' ') {s=String(s); while(s.length<l){s=c+s} return s}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, isObj: function (o) {return !(o == null || typeof o != 'object' || o.constructor.name !== 'Object')}}

var dirtree

import kxk from "../../../kxk.js"
let kutil = kxk.kutil
let kseg = kxk.kseg
let kstr = kxk.kstr
let slash = kxk.slash
let post = kxk.post

import nfs from "../../../kxk/nfs.js"

import prjcts from "../../index/prjcts.js"

import theme from "../../theme/theme.js"
import icons from "../../theme/icons.js"

import git from "../../util/git.js"

import choices from "../menu/choices.js"

import diritem from "./diritem.js"

import rgxs from './quicky.json' with { type : "json" }

dirtree = (function ()
{
    _k_.extend(dirtree, choices)
    function dirtree (screen, name, features)
    {
        this["dirItems"] = this["dirItems"].bind(this)
        this["setState"] = this["setState"].bind(this)
        this["onSessionMerge"] = this["onSessionMerge"].bind(this)
        this["setRoot"] = this["setRoot"].bind(this)
        this["onFileChange"] = this["onFileChange"].bind(this)
        this["onGitStatus"] = this["onGitStatus"].bind(this)
        dirtree.__super__.constructor.call(this,screen,name,features)
        this.state.syntax.setRgxs(rgxs)
        post.on('session.merge',this.onSessionMerge)
        post.on('file.change',this.onFileChange)
        post.on('git.status',this.onGitStatus)
        this.frontRoundOffset = 0
    }

    dirtree.prototype["onGitStatus"] = function (status)
    {
        var item

        var list = _k_.list(this.items)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            item = list[_a_]
            if (_k_.in(item.path,status.changed))
            {
                item.modified = true
                item.tilde = _k_.lpad(item.depth * 4 - 3) + diritem.symbolName(item)
                this.set(this.items,this.currentIndex())
                post.emit('redraw')
            }
        }
    }

    dirtree.prototype["onFileChange"] = function (info)
    {
        var item

        if (info.path.startsWith(this.currentRoot))
        {
            if (info.change === 'rename')
            {
                var list = _k_.list(this.items)
                for (var _a_ = 0; _a_ < list.length; _a_++)
                {
                    item = list[_a_]
                    if (item.path === info.path)
                    {
                        return
                    }
                }
                this.setRoot(this.currentRoot,{redraw:true,index:this.currentIndex()})
            }
            if (_k_.in(info.change,['remove','deleted']))
            {
                var list1 = _k_.list(this.items)
                for (var _b_ = 0; _b_ < list1.length; _b_++)
                {
                    item = list1[_b_]
                    if (item.path === info.path)
                    {
                        this.setRoot(this.currentRoot,{redraw:true,index:this.currentIndex()})
                        return
                    }
                }
            }
        }
    }

    dirtree.prototype["setRoot"] = async function (path, opt)
    {
        var dir, item, items, _88_29_

        opt = (opt != null ? opt : {})
        dir = slash.untilde(path)
        items = await this.dirItems(dir,'dirtree.setRoot')
        this.currentRoot = dir
        if (_k_.empty(items))
        {
            return
        }
        var list = _k_.list(items)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            item = list[_a_]
            item.tilde = ' ' + diritem.symbolName(item)
        }
        items.sort((function (a, b)
        {
            return this.weight(a) - this.weight(b)
        }).bind(this))
        this.set(items,((_88_29_=opt.index) != null ? _88_29_ : 0))
        this.restoreSessionState(opt)
        if (opt.redraw)
        {
            return post.emit('redraw')
        }
    }

    dirtree.prototype["restoreSessionState"] = function (opt)
    {
        var key, state, value

        opt = (opt != null ? opt : {})
        opt.redraw = true
        state = ked_session.get(this.name,{})
        if (_k_.empty(state.open))
        {
            return
        }
        for (key in state.open)
        {
            value = state.open[key]
            if (key.startsWith(this.currentRoot))
            {
                this.openDir(this.itemForPath(key),opt)
            }
        }
    }

    dirtree.prototype["onSessionMerge"] = function (recent)
    {
        return this.setState(recent[this.name])
    }

    dirtree.prototype["setState"] = function (state)
    {
        if (_k_.empty(state))
        {
            return
        }
        if (!(_k_.isObj(state)))
        {
            return
        }
        return ked_session.set(this.name,state)
    }

    dirtree.prototype["emitAction"] = function (action, choice, event)
    {
        if (action === 'hover')
        {
            this.grabFocus()
            if (!_k_.empty(event.mods) && choice.type === 'file')
            {
                post.emit('quicky',choice.path)
            }
            return
        }
        if (action === 'cmd+delete')
        {
            post.emit('file.trash',choice.path)
            return
        }
        switch (choice.type)
        {
            case 'dir':
                switch (action)
                {
                    case 'click':
                    case 'space':
                        if (action === 'click' && event.mods)
                        {
                            return post.emit('dircol.root',choice.path)
                        }
                        if (!choice.open)
                        {
                            this.openDir(choice,{redraw:true})
                        }
                        else
                        {
                            this.closeDir(choice)
                        }
                        return

                    case 'right':
                        if (!choice.open)
                        {
                            this.openDir(choice,{redraw:true})
                        }
                        else
                        {
                            this.selectNextKeepOffset()
                        }
                        return

                    case 'left':
                        if (choice.open)
                        {
                            this.closeDir(choice,{redraw:true})
                        }
                        else
                        {
                            this.selectPrevKeepOffset()
                        }
                        return

                    case 'delete':
                    case 'esc':
                        if (!choice.open)
                        {
                            this.selectOpenSiblingAboveOrParent()
                        }
                        if (choice.open)
                        {
                            this.closeDir(choice,{redraw:true})
                        }
                        return

                    case 'doubleclick':
                    case 'return':
                        return post.emit('dircol.root',choice.path)

                }

                break
            case 'file':
                switch (action)
                {
                    case 'left':
                        return this.selectPrevKeepOffset()

                    case 'right':
                        return this.selectNextKeepOffset()

                    case 'delete':
                    case 'esc':
                        return this.selectParent()

                    case 'drag':
                    case 'space':
                        return post.emit('quicky',choice.path)

                    case 'click':
                    case 'return':
                        return post.emit('file.open',choice.path)

                }

                break
        }

        return dirtree.__super__.emitAction.call(this,action,choice,event)
    }

    dirtree.prototype["openDir"] = async function (dirItem, opt)
    {
        var depth, index, item, items, state, _228_31_, _232_48_, _244_20_, _246_26_

        if (_k_.empty(dirItem))
        {
            return
        }
        if (dirItem.open)
        {
            return
        }
        opt = (opt != null ? opt : {})
        dirItem.open = true
        items = await this.dirItems(dirItem.path,'dirtree.openDir')
        dirItem.tilde = dirItem.tilde.replace(icons.dir_close,icons.dir_open)
        state = ked_session.get(this.name,{})
        depth = (((_228_31_=dirItem.depth) != null ? _228_31_ : 0)) + 1
        var list = _k_.list(items)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            item = list[_a_]
            item.depth = depth
            item.tilde = _k_.lpad(1 + depth * 2) + diritem.symbolName(item)
            if (item.type === 'dir' && (state.open != null ? state.open[item.path] : undefined))
            {
                if (_k_.empty(opt.select) && _k_.empty(opt.index))
                {
                    opt.select = dirItem
                }
                opt.redraw = true
                this.openDir(item,opt)
            }
        }
        items.sort((function (a, b)
        {
            return this.weight(a) - this.weight(b)
        }).bind(this))
        index = this.items.indexOf(dirItem)
        kutil.insert(this.items,index + 1,items)
        if ((opt.index != null))
        {
            index = opt.index
        }
        else if ((opt.select != null))
        {
            index = this.items.indexOf(opt.select)
        }
        this.set(this.items,index)
        ked_session.set(`${this.name}▸open▸${dirItem.path}`,'✔')
        git.status(dirItem.path)
        if (opt.redraw)
        {
            return post.emit('redraw')
        }
    }

    dirtree.prototype["closeDir"] = function (dirItem, opt)
    {
        var index, numChildren

        if (_k_.empty(dirItem))
        {
            return
        }
        opt = (opt != null ? opt : {})
        dirItem.open = false
        dirItem.tilde = dirItem.tilde.replace(icons.dir_open,icons.dir_close)
        index = this.items.indexOf(dirItem)
        numChildren = 0
        while (this.items[index + numChildren + 1].path.startsWith(dirItem.path))
        {
            numChildren += 1
        }
        kutil.replace(this.items,index + 1,numChildren,[])
        this.set(this.items,index)
        ked_session.del(`${this.name}▸open▸${dirItem.path}`)
        if (opt.redraw)
        {
            return post.emit('redraw')
        }
    }

    dirtree.prototype["dirItems"] = async function (dir, info)
    {
        var items

        try
        {
            return items = await nfs.list(dir,{recursive:false})
        }
        catch (err)
        {
            console.error(`list error -- ${info}`,err)
            return
        }
    }

    dirtree.prototype["set"] = function (items, index = 0)
    {
        var oldTop

        oldTop = this.state.s.view[1]
        dirtree.__super__.set.call(this,items,'tilde')
        this.state.setView([0,oldTop])
        this.state.selectLine(index)
        return this.state.setMainCursor(0,index)
    }

    dirtree.prototype["selectPrevKeepOffset"] = function ()
    {
        this.selectPrev()
        if (this.current().type === 'file')
        {
            post.emit('quicky',this.current().path)
        }
        return this.state.setView([0,this.state.s.view[1] - 1])
    }

    dirtree.prototype["selectNextKeepOffset"] = function ()
    {
        if (this.current().type === 'file' && ked_session.get('editor▸file') !== this.current().path)
        {
            post.emit('quicky',this.current().path)
            return
        }
        this.selectNext()
        if (this.current().type === 'file')
        {
            post.emit('quicky',this.current().path)
        }
        return this.state.setView([0,this.state.s.view[1] + 1])
    }

    dirtree.prototype["selectOpenSiblingAboveOrParent"] = function ()
    {
        var index

        index = this.fuzzied.indexOf(this.current())
        index -= 1
        while (index > 0 && this.fuzzied[index].depth >= this.current().depth && !this.fuzzied[index].open)
        {
            index -= 1
        }
        this.state.selectLine(index)
        return this.state.setMainCursor(0,index)
    }

    dirtree.prototype["selectParent"] = function ()
    {
        var index

        index = this.fuzzied.indexOf(this.current())
        index -= 1
        while (index > 0 && this.fuzzied[index].depth >= this.current().depth)
        {
            index -= 1
        }
        this.state.selectLine(index)
        return this.state.setMainCursor(0,index)
    }

    dirtree.prototype["drawSelections"] = function ()
    {
        var bg, li, x, xs, y

        if (li = this.indexOfOpenFile())
        {
            bg = theme.gutter.bg
            y = li - this.state.s.view[1]
            if (y < this.cells.rows && li < this.state.s.lines.length)
            {
                xs = kseg.headCount(this.state.s.lines[li],' ')
                this.cells.set(xs - 1 - this.state.s.view[0],y,'',bg,this.color.bg)
                for (var _a_ = x = xs, _b_ = this.cells.cols; (_a_ <= _b_ ? x < this.cells.cols : x > this.cells.cols); (_a_ <= _b_ ? ++x : --x))
                {
                    this.cells.set_bg(x - this.state.s.view[0],y,bg)
                }
            }
        }
        return dirtree.__super__.drawSelections.call(this)
    }

    dirtree.prototype["indexOfOpenFile"] = function ()
    {
        var idx, item, _377_45_

        if (!(global.ked_editor_file != null))
        {
            return
        }
        var list = _k_.list(this.fuzzied)
        for (idx = 0; idx < list.length; idx++)
        {
            item = list[idx]
            if (item.path === ked_editor_file)
            {
                return idx
            }
        }
    }

    dirtree.prototype["itemForPath"] = function (p)
    {
        var idx, item

        var list = _k_.list(this.items)
        for (idx = 0; idx < list.length; idx++)
        {
            item = list[idx]
            if (slash.samePath(item.path,p))
            {
                return item
            }
        }
    }

    dirtree.prototype["itemIndexForPath"] = function (p)
    {
        var idx, item

        var list = _k_.list(this.items)
        for (idx = 0; idx < list.length; idx++)
        {
            item = list[idx]
            if (slash.samePath(item.path,p))
            {
                return idx
            }
        }
    }

    dirtree.prototype["weight"] = function (item)
    {
        var p, w

        p = slash.parse(item.path)
        w = 0
        if (item.type === 'file')
        {
            w += 10000
        }
        w += kstr.weight(p.file)
        return w
    }

    return dirtree
})()

export default dirtree;

File: kakao.app/love/lua/view/colmns/dirtree.lua
--[[
0000000    000  00000000   000000000  00000000   00000000  00000000
000   000  000  000   000     000     000   000  000       000     
000   000  000  0000000       000     0000000    0000000   0000000 
000   000  000  000   000     000     000   000  000       000     
0000000    000  000   000     000     000   000  00000000  00000000
--]]

-- import rgxs from './quicky.json' with { type: "json" }

diritem = require "view.colmns.diritem"
choices = require "view.menu.choices"


local dirtree = class("dirtree", choices)
    


function dirtree:init(name, features) 
        choices.init(self, name, features)
        
        -- @state.syntax.setRgxs rgxs
        
        post:on('session.merge', self.onSessionMerge)
        post:on('file.change', self.onFileChange)
        post:on('git.status', self.onGitStatus)
        
        self.frontRoundOffset = 0
        return self
    end

--  ███████   ███  █████████         ███████  █████████   ███████   █████████  ███   ███   ███████
-- ███        ███     ███           ███          ███     ███   ███     ███     ███   ███  ███     
-- ███  ████  ███     ███           ███████      ███     █████████     ███     ███   ███  ███████ 
-- ███   ███  ███     ███                ███     ███     ███   ███     ███     ███   ███       ███
--  ███████   ███     ███           ███████      ███     ███   ███     ███      ███████   ███████ 


function dirtree:tilde(item) 
    item.tilde = kstr.lpad(((item.depth * 2) + 1)) .. diritem.symbolName(item)
    return item.tilde
    end


function dirtree:onGitStatus(status) 
        -- todo: handle added files
        
        -- log 'dirtree.onGitStatus' status
        
        for _, item in ipairs(self.items) do 
            if (item.type == 'dir') then 
                local foundChange = false
                
                if status.dirs:has(item.path) then 
                    for _, file in ipairs(status.files) do 
                        if file.startsWith(item.path) then 
                            item.modified = true
                            self:tilde(item)
                            local redraw = true
                            foundChange = true
                            break
                        end
                    end
                end
                
                if (item.modified and not foundChange) then 
                    item.modified = nil
                    self:tilde(item)
                    local redraw = true
                end
            else 
                if status.changed:has(item.path) then 
                    item.modified = true
                    self:tilde(item)
                    local redraw = true
                end
                
                if status.added:has(item.path) then 
                    item.added = true
                    self:tilde(item)
                    local redraw = true
                end
                
                if ((item.modified and not status.changed:has(item.path)) or (item.added and not status.added:has(item.path))) then 
                    item.modified = nil
                    self:tilde(item)
                    local redraw = true
                end
            end
        end
        
        if redraw then 
            self:set(self.items, self:currentIndex())
            return post:emit('redraw')
        end
    end

-- ████████  ███  ███      ████████         ███████  ███   ███   ███████   ███   ███   ███████   ████████
-- ███       ███  ███      ███             ███       ███   ███  ███   ███  ████  ███  ███        ███     
-- ██████    ███  ███      ███████         ███       █████████  █████████  ███ █ ███  ███  ████  ███████ 
-- ███       ███  ███      ███             ███       ███   ███  ███   ███  ███  ████  ███   ███  ███     
-- ███       ███  ███████  ████████         ███████  ███   ███  ███   ███  ███   ███   ███████   ████████


function dirtree:onFileChange(info) 
        if kstr.startWith(info.path, self.currentRoot) then 
            if (info.change == 'rename') then 
                for _, item in ipairs(self.items) do 
                    if (item.path == info.path) then 
                        return
                    end
                end
                
                self:setRoot(self.currentRoot, {index = self.currentIndex()})
            end
            
            if array('remove', 'deleted'):has(info.change) then 
                for _, item in ipairs(self.items) do 
                    if (item.path == info.path) then 
                        self:setRoot(self.currentRoot, {index = self.currentIndex()})
                        return
                    end
                end
            end
        end
    end

-- 00000000    0000000    0000000   000000000  
-- 000   000  000   000  000   000     000     
-- 0000000    000   000  000   000     000     
-- 000   000  000   000  000   000     000     
-- 000   000   0000000    0000000      000     


function dirtree:setRoot(path, opt) 
        opt = opt or ({})
        
        local dir = slash.untilde(path)
        
        local items = self:dirItems(dir, 'dirtree.setRoot')
        
        self.currentRoot = dir
        
        if empty(items) then return end
        
        for item in items:each() do 
            item.depth = 0
            self:tilde(item)
        end
        
        -- items∙sort((a b) -> @weight(a) - @weight(b))
        
        self:set(items, (opt.index or 1))
        
        return self:restoreSessionState(opt)
    end

-- 00000000   00000000   0000000  000000000   0000000   00000000   00000000  
-- 000   000  000       000          000     000   000  000   000  000       
-- 0000000    0000000   0000000      000     000   000  0000000    0000000   
-- 000   000  000            000     000     000   000  000   000  000       
-- 000   000  00000000  0000000      000      0000000   000   000  00000000  


function dirtree:restoreSessionState(opt) 
        opt = opt or ({})
        opt.redraw = true
        
        local state = ked_session:get(self.name, {})
        
        if empty(state.open) then return end
        
        for key, value in pairs(state.open) do 
            if kstr.startsWith(key, self.currentRoot) then 
                self:openDir(self:itemForPath(key), opt)
            end
        end
    end

--  0000000  000000000   0000000   000000000  00000000  
-- 000          000     000   000     000     000       
-- 0000000      000     000000000     000     0000000   
--      000     000     000   000     000     000       
-- 0000000      000     000   000     000     00000000  


function dirtree:onSessionMerge(recent) 
    return self:setState(recent[self.name])
    end


function dirtree:setState(state) 
        if empty(state) then return end
        
        return ked_session:set(self.name, state)
    end

--  0000000    0000000  000000000  000   0000000   000   000  
-- 000   000  000          000     000  000   000  0000  000  
-- 000000000  000          000     000  000   000  000 0 000  
-- 000   000  000          000     000  000   000  000  0000  
-- 000   000   0000000     000     000   0000000   000   000  


function dirtree:emitAction(action, choice, event) 
        print("DIRTREE emitAction", action)
        if (action == 'hover') then 
            self:grabFocus()
            if ((event.alt or event.cmd) and (choice.type == 'file')) then 
                post:emit('quicky', choice.path)
            end
            
            return
        end
        
        if (action == 'cmd+delete') then 
            if (choice.type == 'file') then 
                post:emit('file.trash', choice.path)
            end
            
            return
        end
        
        if (choice.type == 'dir') then 
                if (action == 'click') or (action == 'space') then 
                        if ((action == 'click') and event.mods) then return post:emit('dircol.root', choice.path) end
                        
                        if not choice.open then self:openDir(choice)
                        else self:closeDir(choice)
                        end
                        
                        return
                elseif (action == 'right') then 
                        if not choice.open then self:openDir(choice)
                        else self:selectNextKeepOffset()
                        end
                        
                        return
                elseif (action == 'left') then 
                        if choice.open then self:closeDir(choice)
                        else self:selectPrevKeepOffset()
                        end
                        
                        return
                elseif (action == 'delete') or (action == 'esc') then 
                        if not choice.open then self:selectOpenSiblingAboveOrParent() end
                        if choice.open then self:closeDir(choice) end
                        return
                elseif (action == 'doubleclick') or (action == 'return') then return post:emit('dircol.root', choice.path)
                end
        elseif (choice.type == 'file') then 
                if (action == 'left') then return self:selectPrevKeepOffset()
                elseif (action == 'right') then return self:selectNextKeepOffset()
                elseif (action == 'delete') or (action == 'esc') then return self:selectParent()
                elseif (action == 'drag') or (action == 'space') then return post:emit('quicky', choice.path)
                elseif (action == 'click') or (action == 'return') then return post:emit('file.open', choice.path)
                end
        end
        
        return choices.emitAction(self, action, choice, event)
    end

--  0000000   00000000   00000000  000   000  0000000    000  00000000   
-- 000   000  000   000  000       0000  000  000   000  000  000   000  
-- 000   000  00000000   0000000   000 0 000  000   000  000  0000000    
-- 000   000  000        000       000  0000  000   000  000  000   000  
--  0000000   000        00000000  000   000  0000000    000  000   000  


function dirtree:openDir(dirItem, opt) 
        if empty(dirItem) then return end
        if dirItem.open then return end
        
        opt = opt or ({})
        
        dirItem.open = true
        
        local items = self:dirItems(dirItem.path, 'dirtree.openDir')
        
        dirItem.tilde = string.gsub(dirItem.tilde, icons.dir_close, icons.dir_open)
        
        local state = ked_session:get(self.name, {})
        
        local depth = (dirItem.depth + 1)
        for _, item in ipairs(items) do 
            item.depth = depth
            self:tilde(item)
            if state.open then 
                if ((item.type == 'dir') and state.open[item.path]) then 
                    if empty((opt.select and empty), opt.index) then 
                        opt.select = dirItem
                    end
                    
                    opt.redraw = true
                    self:openDir(item, opt)
                end
            end
        end
        
        items:sort(function (a, b) 
    return (self:weight(a) < self:weight(b))
end)
        
        local index = self.items:findWith(function (i) 
    return (i.path == dirItem.path)
end)
        
        print("SPLICE INDEX", index)
        self.items:splice((index + 1), 0, unpack(items))
        
        if opt.index then 
            index = opt.index
        elseif opt.select then 
            index = self.items:indexof(opt.select)
        end
        
        self:set(self.items, index)
        
        ked_session:set("" .. tostring(self.name) .. "▸open▸" .. tostring(dirItem.path) .. "", '✔')
        
        -- git.status dirItem.path
        return self
    end

--  0000000  000       0000000    0000000  00000000  0000000    000  00000000   
-- 000       000      000   000  000       000       000   000  000  000   000  
-- 000       000      000   000  0000000   0000000   000   000  000  0000000    
-- 000       000      000   000       000  000       000   000  000  000   000  
--  0000000  0000000   0000000   0000000   00000000  0000000    000  000   000  


function dirtree:closeDir(dirItem, opt) 
        if empty(dirItem) then return end
        
        opt = opt or ({})
        
        dirItem.open = false
        
        dirItem.tilde = string.gsub(dirItem.tilde, icons.dir_open, icons.dir_close)
        
        local index = self.items:findWith(function (i) 
    return (i.path == dirItem.path)
end)
        
        local numChildren = 0
        while ((((index + numChildren) + 1) < self.items:len()) and kstr.startsWith(self.items[((index + numChildren) + 1)].path, dirItem.path)) do 
            numChildren = numChildren + 1
        end
        
        self.items:splice((index + 1), numChildren)
        
        self:set(self.items, index)
        
        ked_session:del("" .. tostring(self.name) .. "▸open▸" .. tostring(dirItem.path) .. "")
        return self
    end

-- 000  000000000  00000000  00     00   0000000  
-- 000     000     000       000   000  000       
-- 000     000     0000000   000000000  0000000   
-- 000     000     000       000 0 000       000  
-- 000     000     00000000  000   000  0000000   


function dirtree:dirItems(dir, info) 
        return slash.walk(dir, {recursive = false})
    end

--  0000000  00000000  000000000  
-- 000       000          000     
-- 0000000   0000000      000     
--      000  000          000     
-- 0000000   00000000     000     


function dirtree:set(items, index) 
        index = index or 1
        
        local oldTop = self.state.s.view[2]
        choices.set(self, items, 'tilde')
        self.state:setView(array(1, oldTop))
        self.state:selectLine(index)
        return self.state:setMainCursor(1, index)
    end

--  0000000  00000000  000      00000000   0000000  000000000  
-- 000       000       000      000       000          000     
-- 0000000   0000000   000      0000000   000          000     
--      000  000       000      000       000          000     
-- 0000000   00000000  0000000  00000000   0000000     000     


function dirtree:selectPrevKeepOffset() 
        self:selectPrev()
        if (self:current().type == 'file') then 
            post:emit('quicky', self:current().path)
        end
        
        return self.state:setView(array(1, (self.state.s.view[1] - 1)))
    end


function dirtree:selectNextKeepOffset() 
        if ((self:current().type == 'file') and (ked_session:get('editor▸file') ~= self:current().path)) then 
            post:emit('quicky', self:current().path)
            return
        end
        
        self:selectNext()
        if (self:current().type == 'file') then 
            post:emit('quicky', self:current().path)
        end
        
        return self.state:setView(array(1, (self.state.s.view[1] + 1)))
    end


function dirtree:selectOpenSiblingAboveOrParent() 
        local index = self.fuzzied.indexOf(self:current())
        index = index - 1
        while (((index > 0) and (self.fuzzied[index].depth >= self:current().depth)) and not self.fuzzied[index].open) do 
            index = index - 1
        end
        
        self.state.selectLine(index)
        return self.state.setMainCursor(1, index)
    end


function dirtree:selectParent() 
        local index = self.fuzzied.indexOf(self:current())
        index = index - 1
        while ((index > 0) and (self.fuzzied[index].depth >= self:current().depth)) do 
            index = index - 1
        end
        
        self.state.selectLine(index)
        return self.state.setMainCursor(1, index)
    end


function dirtree:drawSelections() 
        local li = self:indexOfOpenFile()
        if li then 
            local bg = theme.gutter.bg
            local y = (li - self.state.s.view[2])
            if ((y < self.cells.rows) and (li < self.state.s.lines:len())) then 
                local xs = kseg.headCount(self.state.s.lines[li], ' ')
                self.cells:set(((xs - 1) - self.state.s.view[1]), y, '', bg, self.color.bg)
                for x = xs, self.cells.cols-1 do 
                    self.cells:set_bg((x - self.state.s.view[1]), y, bg)
                end
            end
        end
        
        return choices.drawSelections(self)
    end


function dirtree:indexOfOpenFile() 
        local currentFile = ked_session:get('editor▸file')
        
        if empty(currentFile) then return end
        
        for idx, item in ipairs(self.fuzzied) do 
            if (item.path == currentFile) then return idx end
        end
    end


function dirtree:itemForPath(p) 
        for idx, item in ipairs(self.items) do 
            if slash.samePath(item.path, p) then return item end
        end
    end


function dirtree:itemIndexForPath(p) 
        for idx, item in ipairs(self.items) do 
            if slash.samePath(item.path, p) then return idx end
        end
    end

-- 000   000  00000000  000   0000000   000   000  000000000  
-- 000 0 000  000       000  000        000   000     000     
-- 000000000  0000000   000  000  0000  000000000     000     
-- 000   000  000       000  000   000  000   000     000     
-- 00     00  00000000  000   0000000   000   000     000     


function dirtree:weight(item) 
        local p = slash.parse(item.path)
        
        local w = 0
        if (item.type == 'file') then 
            w = w + 10000
        end
        
        -- w += kstr.weight(p.file)
        return w
    end

return dirtree